[AOTInductor] Allow user to explicitly specify Device to run on

SherlockNoMad opened the pull request:
Summary:
AOTInductor currently infer cuda device index by `cudaGetDevice()`. This assumes outer runtime calls `cudaSetDevice()` somewhere, before invoking AOTInductor run. 

This diff adds an explicit argument for specifying target Device. e.g. compiled on "cuda:0", run on "cuda:1". 



todo:
- Are the changes in interface.h BC breaking? as it changes the function signatures in .so file. Might just need introduce a new "Create" function.

Test Plan: CI

Differential Revision:
D52747132

Privacy Context Container: 368960445142440




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

chenyang78 reviewed (approved):
LGTM. Thanks!

chenyang78 commented:
Please fix the lint failure. Thanks.

desertfire commented on torch/csrc/inductor/aoti_runtime/interface.h:
```diff
@@ -52,6 +54,15 @@ AOTIRuntimeError AOTInductorModelContainerCreate(
+AOTIRuntimeError AOTInductorModelContainerCreateWithDevice(
+    AOTInductorModelContainerHandle* container_handle,
+    size_t num_models,
+    const char* device_str,
```
What happens if `device_str == nullptr`? Let's document the behavior by adding comments.

SherlockNoMad replied:
commented. 
device_str cannot be nullptr, and it must be a valid device str.

desertfire commented on torch/csrc/inductor/aoti_runner/model_container_runner_cuda.cpp:
```diff
@@ -7,7 +7,7 @@ AOTIModelContainerRunnerCuda::AOTIModelContainerRunnerCuda(
     const std::string& model_so_path,
```
Should we also update the signature of `AOTIModelContainerRunnerCuda`?

SherlockNoMad replied:
done.

SherlockNoMad committed:
[AOTInductor] Allow user to explicitly specify Device to run on (#117413)

Summary:

AOTInductor currently infer cuda device index by `cudaGetDevice()`. This assumes outer runtime calls `cudaSetDevice()` somewhere, before invoking AOTInductor run. 

This diff adds an explicit argument for specifying target Device. e.g. compiled on "cuda:0", run on "cuda:1". 



todo:
- Are the changes in interface.h BC breaking? as it changes the function signatures in .so file. Might just need introduce a new "Create" function.

Test Plan: CI

Reviewed By: chenyang78

Differential Revision: D52747132
